### PR TITLE
fix acl builds on ubuntu

### PR DESCRIPTION
**Pull request by cosmicexplorer**
### Problem

Very strangely, every single version of the `acl` package started to fail builds on my personal ubuntu laptop recently. **I'm not sure why this is working for anyone else**, but I was able to identify three separate issues:
1. `self.build_system_flags(...)` was failing to inject `-lintl` for some reason.
2. Versions up to `2.2.52` didn't have any `./autogen.sh`, but still required a special incantation instead of just running `autoreconf`.
3. Versions up to `2.2.51` required a specific header file from an older version of the `attr` dependency.

### Solution
1. Use `self.inject_flags(...)` to inject `-lintl` **(I have no clue why this works!)**.
2. Patch the codebase's `Makefile` to use their special `autoconf` incantation for versions up to `2.2.52`.
3. Pin the `attr` dependency to `2.4.47` for versions up to `2.2.51` *(older versions of `attr` do not work either, it must be exactly this version!)*.

### Result
All versions of the `acl` package now build on my laptop!

### CODE REVIEW

**Comment by cosmicexplorer** [comment]
Reviewers or bystanders, please feel free to chime in if you have ideas about test cases I should add. It's difficult for me to figure out how to do that because I haven't been able to figure out why this started failing on my machine.

File: var/spack/repos/builtin/packages/attr/package.py [hunk 1]
```diff
@@ -29,8 +34,10 @@ def configure_args(self):
         return args
 
     # Ref. https://www.linuxfromscratch.org/blfs/view/7.5/postlfs/attr.html
+    @when("@2.4.48:")
```
Comment (becker33) [inline var/spack/repos/builtin/packages/attr/package.py]: This method should be unnecessary (since it's not overriding the default behavior, and the other method that does override is gated on a `when` decorator.

File: lib/spack/docs/build_systems/autotoolspackage.rst [hunk 2]
```diff
@@ -138,14 +138,15 @@ dependency, since ``autoconf`` already depends on it.
...
+In some cases, packages will provide a custom script to generate ``configure``,
+often named ``autogen.sh``. The default implementation of the autoreconf phase
+can be overridden to execute such a script.
 
 .. code-block:: python
 
    def autoreconf(self, spec, prefix):
-       which('bash')('autogen.sh')
+       autogen = Executable("./autogen.sh")
+       autogen()
```
Comment (becker33) [inline lib/spack/docs/build_systems/autotoolspackage.rst]: This documentation change seems orthogonal, I'd love to run it by whoever originally put in the shell requirement.

File: var/spack/repos/builtin/packages/acl/package.py [hunk 3]
```diff
@@ -23,14 +27,41 @@ class Acl(AutotoolsPackage):
...
+    def autoreconf(self, spec, prefix):
+        make("autoconf")
+
+    # The "make configure" target in the unpatched Makefile touched this file after executing the
+    # hardcoded ./configure command line. The purpose of this file is not documented, and this step
+    # may not be necessary.
+    @when("@:2.2.52")
+    @run_after("configure")
+    def touch_census(self):
+        touch(".census")
```
Comment (becker33) [inline var/spack/repos/builtin/packages/acl/package.py]: I'm pretty sure this is just to get the targets playing nicely in make, we shouldn't need it when we're not using `make configure` anymore. When you're running `configure`, there's not a file you know will be created by the configure script that you can key on to know whether it ran, but when you run `autoconf` you can key off the existence of the `configure` script. Nothing in the makefile appears to need it, so we should be fine getting rid of this

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/docs/build_systems/autotoolspackage.rst (hunk 2 above)
Comment (cosmicexplorer) [reply lib/spack/docs/build_systems/autotoolspackage.rst]: You're absolutely right. I'll split it into a separate PR.

File: var/spack/repos/builtin/packages/acl/package.py (hunk 3 above)
Comment (cosmicexplorer) [reply var/spack/repos/builtin/packages/acl/package.py]: Thank you, that makes perfect sense!

File: var/spack/repos/builtin/packages/attr/package.py (hunk 1 above)
Comment (cosmicexplorer) [reply var/spack/repos/builtin/packages/attr/package.py]: Fixed!

File: lib/spack/docs/build_systems/autotoolspackage.rst (hunk 2 above)
Comment (cosmicexplorer) [reply lib/spack/docs/build_systems/autotoolspackage.rst]: Hmmm, I actually think the existing code is more correct, and we should not be splitting this out at all. I have changed this PR to use `which('bash')("./autogen.sh")` as well.

File: var/spack/repos/builtin/packages/attr/package.py [hunk 4]
```diff
@@ -23,14 +28,14 @@ def url_for_version(self, version):
...
-    # Ref. https://www.linuxfromscratch.org/blfs/view/7.5/postlfs/attr.html
+    @when("@:2.4.47")
     def install(self, spec, prefix):
-        if self.version >= Version("2.4.48"):
-            make("install")
-        else:
-            make("install", "install-dev", "install-lib")
+        make("install", "install-dev", "install-lib")
+
+        make("install", "install-dev", "install-lib")
```
Comment (alecbcs) [inline var/spack/repos/builtin/packages/attr/package.py]: @cosmicexplorer is this a typo? These looks to both be the same line to me.
```suggestion
        make("install", "install-dev", "install-lib")
```

Comment (alecbcs) [reply var/spack/repos/builtin/packages/attr/package.py]: ping @cosmicexplorer

Comment (alecbcs) [reply var/spack/repos/builtin/packages/attr/package.py]: ping @cosmicexplorer

Comment (cosmicexplorer) [reply var/spack/repos/builtin/packages/attr/package.py]: Yes! This was no longer necessary after some of the earlier fixes.

File: var/spack/repos/builtin/packages/acl/package.py [hunk 5]
```diff
@@ -23,14 +27,33 @@ class Acl(AutotoolsPackage):
...
 
+    depends_on("attr")
+    # The configure step fails to find attr/xattr.h for these versions of acl unless this exact
+    # version of attr is used.
+    depends_on("attr@2.4.47", when="@:2.2.51")
+
+    # In 2.2.53 and later, the codebase provides a nice autogen script.
+    @when("@2.2.53:")
+    def autoreconf(self, spec, prefix):
+        which("bash")("./autogen.sh")
```
Comment (alecbcs) [inline var/spack/repos/builtin/packages/acl/package.py:41]:
```suggestion
    def autoreconf(self, spec, prefix):
        autogen = Executable("./autogen.sh")
        autogen()
```
Acl's `autogen.sh` doesn't make use of any bash-isms so we can launch this instead with a standard `sh` script which is more compatible for folks using distributions like Alpine.

File: var/spack/repos/builtin/packages/acl/package.py [hunk 6]
```diff
@@ -23,14 +27,33 @@ class Acl(AutotoolsPackage):
     depends_on("autoconf", type="build")
     depends_on("automake", type="build")
     depends_on("libtool", type="build")
-    depends_on("attr")
+    depends_on("bash", type="build")
```
Comment (alecbcs) [inline var/spack/repos/builtin/packages/acl/package.py:30]:
```suggestion
```

We can remove the dependency on bash if we instead execute the `./autogen.sh` script directly.

**Review by alecbcs — CHANGES_REQUESTED** [review]
A couple of minor suggestions to reduce additional dependencies.

File: var/spack/repos/builtin/packages/attr/package.py [hunk 7]
```diff
@@ -12,6 +12,11 @@ class Attr(AutotoolsPackage):
     homepage = "https://savannah.nongnu.org/projects/attr"
     url = "http://download.savannah.gnu.org/releases/attr/attr-2.4.47.src.tar.gz"
 
+    maintainers("cosmicexplorer")
+
+    # FIXME: spack does not validate any of the checksums for this package for some reason!
```
Comment (alecbcs) [inline var/spack/repos/builtin/packages/attr/package.py:17]:
```
> spack checksum --verify attr 2.4.48 2.4.47 2.4.46
```

All seem to work for me. What do you mean by Spack not validating the checksums?

**Comment by alecbcs** [comment]
ping @cosmicexplorer

**Comment by alecbcs** [comment]
ping @cosmicexplorer

**Comment by alecbcs** [comment]
@cosmicexplorer going to go ahead and close this PR since I haven't heard back. Testing acl recently I'm not seeing a build failure on Ubuntu. Happy to reopen if you've got time and would like to continue this PR.
